File: Application/Api/Products/List/Handlers/ListProductsHandler.cs
using Application.ErrorHandling.Application;
using Application.Interfaces.Persistence;
using MediatR;
using OneOf;

namespace Application.Api.Products.List.Handlers;

public class ListProductsHandler : IRequestHandler<ListProductsQuery, OneOf<ListProductsResult, List<PlainApplicationError>>>
{
    private readonly IProductRepository _productRepository;

    public ListProductsHandler(IProductRepository productRepository)
    {
        _productRepository = productRepository;
    }

    public async Task<OneOf<ListProductsResult, List<PlainApplicationError>>> Handle(ListProductsQuery request, CancellationToken cancellationToken)
    {
        var products = await _productRepository.FindAllAsync(
            name: request.Name,
            minPrice: request.MinPrice,
            maxPrice: request.MaxPrice,
            description: request.Description,
            createdBefore: request.CreatedBefore,
            createdAfter: request.CreatedAfter
        );

        var result = new ListProductsResult(products: products);
        return result;
    }
}

File: frontend/src/application/commands/products/listProducts/ListProductsHandler.ts
import IProductDataAccess from "../../../interfaces/dataAccess/IProductDataAccess";
import { ICommandHandler } from "../../ICommandHandler";
import { err, ok } from "neverthrow";
import IListProductsCommand from "./ListProductsCommand";
import IListProductsResult from "./IListProductsResult";

export default class ListProductsHandler implements ICommandHandler<IListProductsCommand, IListProductsResult> {
    private _productDataAccess: IProductDataAccess;

    constructor(props: { productDataAccess: IProductDataAccess }) {
        this._productDataAccess = props.productDataAccess;
    }

    async handle(request: IListProductsCommand): Promise<IListProductsResult> {
        try {
            const result = await this._productDataAccess.listProducts({
                id: request.id,
                name: request.name,
                minPrice: request.minPrice,
                maxPrice: request.maxPrice,
                description: request.description,
                createdBefore: request.createdBefore,
                createdAfter: request.createdAfter,
                orderBy: request.orderBy
            });

            if (result.isErr()) {
                return err({ type: "API", data: result.error });
            }

            return ok({ products: result.value });
        } catch (error: unknown) {
            return err({ type: "Exception", data: error });
        }
    }
}
